File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root == None:
            return []
        
        values = [[root.val]]
        queue = [(root, 0)]
        
        while queue:
            node, level = queue.pop(0)
            if node == None:
                continue
                
            child_vals = []
            if node.left != None:
                child_vals.append(node.left.val)
            if node.right != None:
                child_vals.append(node.right.val)
                
            new_level = level+1
            if len(child_vals) != 0:
                if len(values) < new_level + 1:
                    values.append(child_vals)
                else:    
                    values[new_level].extend(child_vals)    
                
            queue.append((node.left, new_level))
            queue.append((node.right, new_level))
        
        return values

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNodeWithLevel struct {
    *TreeNode
    Level int
}

func levelOrder(root *TreeNode) [][]int {
    if root == nil {
        return nil
    }
    
    queue := []*TreeNodeWithLevel{
        { root, 0 },
    }
    values := [][]int{}
    
    for len(queue) > 0 {
        node := queue[0]
        queue = queue[1:]
        
        if node.TreeNode == nil {
            continue
        }
        
        if len(values) < node.Level+1 {
            values = append(values, []int{node.Val})
        } else {
            values[node.Level] = append(values[node.Level], node.Val)
        }
        
        queue = append(queue, &TreeNodeWithLevel{
            node.Left,
            node.Level + 1,
        })
        queue = append(queue, &TreeNodeWithLevel{
            node.Right,
            node.Level + 1,
        })
    }
    
    return values
}